nixos/modules/system/boot/stratisroot.nix

```diff
@@ -0,0 +1,76 @@
+{ config, lib, pkgs, ... }:
+let
+  types = lib.types;
+in
+{
+  options.boot.stratis = {
+    rootPoolUuid = lib.mkOption {
+      type = types.uniq types.str;
+      description = lib.mdoc ''
+        UUID of the stratis pool that the root fs is located in
+      '';
+      example = "04c68063-90a5-4235-b9dd-6180098a20d9";
+    };
+  };
+  config = {
+    assertions = [
+      {
+        assertion = config.boot.initrd.systemd.enable;
+        message = "stratis root fs requires systemd initrd";
+      }
+      {
+        assertion = lib.strings.hasPrefix "/dev/stratis/" config.fileSystems."/".device;
+        message = "root file system must live in /dev/stratis";
+      }
+      {
+        assertion = config.boot.plymouth.enable;
+        message = "plymouth must be enabled for stratis pool password prompt";
+      }
+    ];
+    boot.initrd = {
+      systemd = {
+        storePaths = [
+          "${pkgs.stratisd}/lib/udev/stratis-base32-decode"
+          "${pkgs.stratisd}/lib/udev/stratis-str-cmp"
+          "${pkgs.lvm2.bin}/bin/dmsetup"
+          "${pkgs.stratisd}/libexec/stratisd-min"
+          "${pkgs.stratisd.initrd}/bin/stratis-rootfs-setup"
+        ];
+        packages = [pkgs.stratisd.initrd];
+        extraBin = {
+          thin_check = "${pkgs."thin-provisioning-tools"}/bin/thin_check";
+          thin_repair = "${pkgs."thin-provisioning-tools"}/bin/thin_repair";
+          thin_metadata_size = "${pkgs."thin-provisioning-tools"}/bin/thin_metadata_size";
+          stratis-min = "${pkgs.stratisd}/bin/stratis-min";
+        };
+        services = {
+          stratis-setup = {
+            description = "setup for Stratis root filesystem";
+            unitConfig.DefaultDependencies = "no";
+            conflicts = [ "shutdown.target" ];
```

Comment:
Should probably include `initrd-switch-root.target`, as well as putting `shutdown.target` and `initrd-switch-root.target` in the `before` clause. It's not really necessary in this case but it's good practice. Basically, it's how you ensure a unit is properly stopped before reaching `initrd-switch-root.target` or `shutdown.target`. Again, doesn't really matter here because it's a oneshot with no `ExecStop`, but it's good practice. e.g. Someone might add an override to the unit that gives it an `ExecStop` for some reason.

Reply:
OK, I've added that